_posts/2020-03-18-how-to-use-agent-information-on-puppetserver.md

```diff
@@ -0,0 +1,166 @@
+
+**Example**
+
+The [mysql module](https://forge.puppet.com/puppetlabs/mysql) has this `exec` resource to clean up after installation. This should never run after `${secret_file}` has been dealt with.
+```puppet
+exec { 'remove install pass':
+  command => $rm_pass_cmd,
+  onlyif  => "test -f ${secret_file}",
+  path    => '/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin'
+}
+```
+(from [root_password.pp](https://github.com/puppetlabs/puppetlabs-mysql/blob/312aca80ba7699ab37f9051d871aac2e05b217b3/manifests/server/root_password.pp#L20-L24))
+
+**Use it for ...**
+
+* running one-off commands on a condition.
+
+**Don't use it if ...**
+
+* an exec is not a good fit for your problem.
+* the condition is very complex or timeconsuming.
+* the condition has to be repeated often.
+
+**Further Reading**
+
+* [Resource Type: `exec`](https://puppet.com/docs/puppet/latest/types/exec.html)
+
+## Custom Facts
+
+**What is it?**
+
+Custom facts are distributed to Puppet clients and are available for use in manifests and templates,
+just like any other fact is.
+They can be used to write conditional expressions based on site-specific data that isn’t available via Facter.
+
+**Example**
+
+The apt module has a few [custom facts](https://github.com/puppetlabs/puppetlabs-apt#facts). The `apt_update_last_success` fact is used in [update.pp](https://github.com/puppetlabs/puppetlabs-apt/blob/master/manifests/update.pp) to decide whether or not it is necessary to update apt caches depending on the various configurations specified.
+
+**Use it ...**
+
+* when you need to centrally collect and evaluate the values.
+* when you want to get the value from all your nodes.
+* to change the set of resources and classes included for each node.
+
+**Don't use it if ...**
+
+* the computed value is sensitive and should not be transmitted or stored outside the node.
+* a simpler solution (see above) works.
+
+**Further Reading**
+
+* [Custom facts](https://puppet.com/docs/puppet/latest/fact_overview.html): the current puppet documentation
+* [Custom facts walkthrough](https://puppet.com/docs/puppet/latest/custom_facts.html): how to write your own facts
+* [Starting out with Puppet custom facts](https://glennsarti.github.io/blog/puppet-ruby-facts/): a quick start to converting external facts to ruby custom facts on Windows
+
+## Custom Resources
+
+**What is it?**
+
+With a native ruby plugin you can fully take control of how puppet manages a resource. Through this you get fine-grained control over reporting and enforcing system state through puppet.
+
+**Example**
+
+The [mysql module](https://forge.puppet.com/puppetlabs/mysql) contains custom types and providers to control databases, users and grants.
+
+**Use it for ...**
+
+* things that can be modeled as a puppet resource.
+* fine-grained control.
+* full integration into the puppet tooling.
+
+**Don't use it if ...**
+
+* you do not want to code ruby.
```

Comment:
Although it's technically true that writing Types/Providers using the resource API still requires Ruby, it is much easier than it used to be. I realize that technically the doc link below mentions it, but I'd suggest calling out the PDK and Resource API specifically as making it easier than the olden days...

Reply:
Thank you :heart: